homeassistant/components/whirlpool/sensor.py

```diff
@@ -0,0 +1,263 @@
+"""The Washer/Dryer Sensor for Whirlpool Appliances."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+import logging
+
+from whirlpool.washerdryer import MachineState, WasherDryer
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_UNKNOWN
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+from homeassistant.util.dt import utcnow
+
+from . import WhirlpoolData
+from .const import DOMAIN
+
+TANK_FILL = {
+    "0": "Unknown",
+    "1": "Empty",
+    "2": "25%",
+    "3": "50%",
+    "4": "100%",
+    "5": "Active",
+}
+
+MACHINE_STATE = {
+    MachineState.Standby: "Standby",
+    MachineState.Setting: "Setting",
+    MachineState.DelayCountdownMode: "Delay Countdown",
+    MachineState.DelayPause: "Delay Paused",
+    MachineState.SmartDelay: "Smart Delay",
+    MachineState.SmartGridPause: "Smart Grid Pause",
+    MachineState.Pause: "Pause",
+    MachineState.RunningMainCycle: "Running Maincycle",
+    MachineState.RunningPostCycle: "Running Postcycle",
+    MachineState.Exceptions: "Exception",
+    MachineState.Complete: "Complete",
+    MachineState.PowerFailure: "Power Failure",
+    MachineState.ServiceDiagnostic: "Service Diagnostic Mode",
+    MachineState.FactoryDiagnostic: "Factory Diagnostic Mode",
+    MachineState.LifeTest: "Life Test",
+    MachineState.CustomerFocusMode: "Customer Focus Mode",
+    MachineState.DemoMode: "Demo Mode",
+    MachineState.HardStopOrError: "Hard Stop or Error",
+    MachineState.SystemInit: "System Initialize",
+}
+
+CYCLE_FUNC = [
+    (WasherDryer.get_cycle_status_filling, "Cycle Filling"),
+    (WasherDryer.get_cycle_status_rinsing, "Cycle Rinsing"),
+    (WasherDryer.get_cycle_status_sensing, "Cycle Sensing"),
+    (WasherDryer.get_cycle_status_soaking, "Cycle Soaking"),
+    (WasherDryer.get_cycle_status_spinning, "Cycle Spinning"),
+    (WasherDryer.get_cycle_status_washing, "Cycle Washing"),
+]
+
+
+ICON_D = "mdi:tumble-dryer"
+ICON_W = "mdi:washing-machine"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def washer_state(washer: WasherDryerClass) -> str | None:
+    """Determine correct states for a washer."""
+
+    if washer.washdry.get_attribute("Cavity_OpStatusDoorOpen") == "1":
+        return "Door open"
+
+    machine_state = washer.washdry.get_machine_state()
+    machine_cycle = None
+
+    for func, cycle_name in CYCLE_FUNC:
+        if func(washer.washdry):
+            machine_cycle = cycle_name
+            break
+
+    if machine_state == MachineState.RunningMainCycle and machine_cycle:
+        return machine_cycle
```

Comment:
I think this can be shortened to:
```suggestion
    if washer.washdry.get_machine_state() == MachineState.RunningMainCycle:
        for func, cycle_name in CYCLE_FUNC:
            if func(washer.washdry):
                return cycle_name
```

Reply:
Nice logic cleanup!  
I did keep machine_state rather than call washer.washdry.get_machine_state() two times.